src/radical/pilot/agent/scheduler/tagging.py

```diff
@@ -0,0 +1,312 @@
+
+__copyright__ = "Copyright 2013-2016, http://radical.rutgers.edu"
+__license__   = "MIT"
+
+
+from .continuous import Continuous
+
+
+# ------------------------------------------------------------------------------
+#
+# This is a simle extension of the Continuous scheduler which makes RP behave
+# like a FiFo: it will only really attempt to schedule units if all older units
+# (units with lower unit ID) have been scheduled.  As such, it relies on unit
+# IDs to be of the format `unit.%d'`.
+#
+class Tagging(Continuous):
+
+    # --------------------------------------------------------------------------
+    #
+    def __init__(self, cfg, session):
+
+        self.slots = None
+        self._tag_history = dict()
+
+        Continuous.__init__(self, cfg, session)
+
+
```

Comment:
This inherits from `Continuous`, but then basically re-implements the core scheduling parts of the parent class.  So any changes in the scheduling algorithm and any bug fixes etc. need to be done in both classes.

Can you think of a way to avoid this?

Reply:
The tagging.py file won't exist, this was just temporary to test the tagging capability. Do you agree with the tagging method I am using right now? Please see the tests under tests/test_da_scheduler/test_tagging_scheduler.py as well

Reply:
Hmm, I see.  The original idea for the RP schedulers though was to keep them lean and mean, and to quickly exchange them as needed / wanted for experiments.  Now the Cont. scheduler seems to grow into a do-it-all thing.  Can't blame you guys, since I started that trend I guess (the continuous scheduler also does scattered allocations...) - but still, I see this as problematic in the long run.  Please do reconsider if you see a way to use inheritance (or some other form of separation) for the tagging functionality.

Reply:
I completely agree. It is becoming bigger and precisely the reason where I kept the tagging in another file. Right now I am committing to the same continuous scheduler file. But let me propose separating two files: Continuous (as it was before), ContinuousDataAware (with tagging).

What do you think @iparask and team, @andre-merzky ?

Reply:
+1

Reply:
Well, `ContinuousBase` could be the old, plain continuous scheduler, but refactored so that it exposes methods which can be reused?  I don't want to pre-scribe that structure though...   

Either way, I'm glad that you are open to the argument.   I share your concern about the deadlines - in the light of that, let me know if I can help with separating those classes...